### PR TITLE
Port swaybar to layer shell

### LINKED ISSUE #552: Text offset in swaybar caused by CJK
When displaying swaybar with some fonts (e.g. the default one, Hack etc.), sway offsets its contents down. Screenshots demonstrating behavior:
2016-03-29-200652_swaygrab
2016-03-29-200659_swaygrab 

It does not happen with some other fonts (like Unifont):
2016-03-29-200626_swaygrab
2016-03-29-200635_swaygrab

My config: https://sicp.me/u/m3FPW.txt

For reference, i3bar exhibits the same behavior:
2016-03-29-191725_2880x1800_scrot
2016-03-29-191732_2880x1800_scrot

(same i3bar config)

### LINKED ISSUE #554: Lack of vertical alignment for swaybar
I noticed bar elements aren't vertically centered:
sway

i3 version for reference:
i3

### LINKED ISSUE #771: Space between swaybar and windows
See screenshot: http://imgur.com/a/IVhAD

Bisected to here:

```
d3752b41405b5572d0ba488aaf969f308e06968a is the first bad commit
commit d3752b41405b5572d0ba488aaf969f308e06968a
Author: David Eklov <david.eklov@gmail.com>
Date:   Mon Jul 4 17:03:55 2016 -0500

    swaybar: Make swaybar a shell surface

:040000 040000 916d52a5a6a03a2a6e99a6c4f3723c92d7c865ec ec0f1d90d309d526316e4ea71c49b24b74ca3ac1 M  swaybar
```

### LINKED ISSUE #1224: swaybar: render workspace buttons at full bar height regardless of font

### LINKED ISSUE #1352: swaybar eats a core with i3status and an external script
When configuring `swaybar` to use [an external script](https://pastebin.com/5EN48VDK) (saved as `~/.config/sway/run-i3status.zsh`) to invoke `i3status`, the `swaybar` process eats a core (CPU usage goes up to ~100/`$(nproc)`) and doesn't display anything (the line is updating too quickly?).

Relevant part of `~/.config/sway/config`:
```
#
# Status Bar:
#
# Read `man 5 sway-bar` for more information about this section.
bar {
    status_command run-i3status.zsh
    position top
    colors {
        statusline #ffffff
        background #323232
        inactive_workspace #32323200 #32323200 #5c5c5c
    }
}
```
See also [`i3status.conf`](https://pastebin.com/BLSjh6cA) and [`/tmp/weather.txt`](https://pastebin.com/Gje17qUH).

Environment information:
```
$ uname -a
Linux archluon 4.12.12-1-ARCH #1 SMP PREEMPT Sun Sep 10 09:41:14 CEST 2017 x86_64 GNU/Linux
$ sway -v
sway version 0.14.0
```

### LINKED ISSUE #1645: Swaybar: Text after a special character is lowered
If I use a special character like '⚡' in my swaybar (in i3status), every character in the specific section are lowered.

screenshot

Line drawn under the text shows a difference in height of one pixel:
screenshot

i3status output: Volume: 65%  |  Wifi: W-Land 62% 117 Mb/s  |  Time: 15:16  |  ⚡ Battery: 60%

[My i3config file](https://gist.githubusercontent.com/timokoesters/bfeea36bd958ed15f8e08235b2332393/raw/05a5d6c8b7fd9b8351f7b9589949b594532bc170/gistfile1.txt)

### PR DESCRIPTION
**Pull request by SirCmpwn**
I overhauled swaybar's rendering stack along the way.

Fixes #1645, fixes #1224, fixes #1352, fixes #771, fixes #554, fixes #552

* [x] swaybar
  * [x] Port to layer surfaces
  * [x] Overhaul rendering
  * [x] IPC
  * [x] text mode status line
* [x] sway
  * [x] bar configuration
  * [x] IPC get_bar_config
  * [x] IPC get_workspaces
  * [x] Emit workspace events
  * [x] Spawn i3bar process
  * [x] Config reloading
  * [x] avoid rearranging all windows all the time (layer shell issue)
  * [x] Rearrange on config reload

Not doing in this patch:

* [ ] hidpi
* [ ] input
* [ ] output hotplugging
* [ ]  i3bar json status line
* [ ] tray

### CODE REVIEW

**Comment by SirCmpwn** [comment]
TODO: figure out what to do with too much text

**Comment by SirCmpwn** [comment]
This is ready for review.

File: common/pango.c [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	pango_layout_set_single_paragraph_mode(layout, 1);
+	pango_layout_set_attributes(layout, attrs);
+	pango_attr_list_unref(attrs);
+	pango_font_description_free(desc);
+	return layout;
+}
+
+void get_text_size(cairo_t *cairo, const char *font, int *width, int *height,
+		int32_t scale, bool markup, const char *fmt, ...) {
+	char *buf = malloc(2048);
```
Comment (emersion) [inline common/pango.c]: Put this on the stack

File: common/pango.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	PangoLayout *layout = get_pango_layout(cairo, font, buf, scale, markup);
+	pango_cairo_update_layout(cairo, layout);
+	pango_layout_get_pixel_size(layout, width, height);
+	g_object_unref(layout);
+	free(buf);
+}
+
+void pango_printf(cairo_t *cairo, const char *font,
+		int32_t scale, bool markup, const char *fmt, ...) {
+	char *buf = malloc(2048);
```
Comment (emersion) [inline common/pango.c]: Same

File: common/pango.c (hunk 1 above)
Comment (SirCmpwn) [reply common/pango.c]: Could also just statically allocate it

File: sway/commands/bar.c [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	}
+
+	// Create new bar with default values
+	struct bar_config *bar = default_bar_config();
+	if (!bar) {
+		return cmd_results_new(CMD_FAILURE, "bar", "Unable to allocate bar state");
+	}
+
+	// set bar id
+	int i;
```
Comment (emersion) [inline sway/commands/bar.c]: This can go inside the loop

File: sway/commands/bar.c [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	for (i = 0; i < config->bars->length; ++i) {
+		if (bar == config->bars->items[i]) {
+			const int len = 5 + numlen(i); // "bar-" + i + \0
+			bar->id = malloc(len * sizeof(char));
+			if (bar->id) {
+				snprintf(bar->id, len, "bar-%d", i);
+			} else {
+				return cmd_results_new(CMD_FAILURE, "bar", "Unable to allocate bar ID");
```
Comment (emersion) [inline sway/commands/bar.c]: `cmd_results_new` allocates stuff

File: common/pango.c (hunk 1 above)
Comment (ascent12) [reply common/pango.c]: If you wanted to be "careful" with stack space, you can use `vsnprintf(NULL, 0, fmt, args)` to get the number of characters to fit the string and use a VLA.

File: sway/commands/bar.c (hunk 4 above)
Comment (SirCmpwn) [reply sway/commands/bar.c]: If you have a crazy bar ID though that could fail and this might not

File: sway/commands/bar/colors.c [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+static struct cmd_results *parse_three_colors(char ***colors, const char *cmd_name, int argc, char **argv) {
+	struct cmd_results *error = NULL;
+	if (argc != 3) {
+		return cmd_results_new(CMD_INVALID, cmd_name, "Requires exactly three color values");
+	}
+
+	int i;
+	for (i = 0; i < 3; i++) {
```
Comment (emersion) [inline sway/commands/bar/colors.c]: `size_t` inside the loop

File: sway/config/output.c [hunk 6]
```diff
@@ -180,19 +180,20 @@ void apply_output_config(struct output_config *oc, swayc_t *output) {
...
-			"./swaybg/swaybg",
-			output_id,
-			oc->background,
-			oc->background_option,
-			NULL,
-		};
+		size_t len = snprintf(NULL, 0, "%s %d %s %s",
+				config->swaybg_command ? config->swaybg_command : "swaybg",
+				output_i, oc->background, oc->background_option);
+		char *command = malloc(len + 1);
```
Comment (emersion) [inline sway/config/output.c]: VLA?

Comment (SirCmpwn) [reply sway/config/output.c]: meh

File: sway/desktop/layer_shell.c [hunk 7]
```diff
@@ -167,11 +166,11 @@ void arrange_layers(struct sway_output *output) {
 			&usable_area, true);
 	arrange_layer(output, &output->layers[ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND],
 			&usable_area, true);
-	memcpy(&output->usable_area, &usable_area, sizeof(struct wlr_box));
```
Comment (emersion) [inline sway/desktop/layer_shell.c]: <3

File: sway/ipc-server.c [hunk 8]
```diff
@@ -357,6 +378,25 @@ void ipc_client_disconnect(struct ipc_client *client) {
 	free(client);
 }
 
+static void ipc_get_workspaces_callback(swayc_t *workspace, void *data) {
+	if (workspace->type == C_WORKSPACE) {
```
Comment (emersion) [inline sway/ipc-server.c]:
```c
if (workspace->type != C_WORKSPACE) {
   return;
}
```

File: swaybar/bar.c [hunk 9]
```diff
@@ -1,390 +1,200 @@
...
-			} else {
-				status_line_mouse_event(&swaybar, x, y, 5);
-			}
-			return;
-		}
-	}
+static void layer_surface_closed(void *_output,
+		struct zwlr_layer_surface_v1 *surface) {
+	// TODO: Deal with hotplugging
+	struct swaybar_output *output = output;
```
Comment (emersion) [inline swaybar/bar.c]: I bet the compiler replaces this with a `nop`

File: swaybar/bar.c [hunk 10]
```diff
@@ -1,390 +1,200 @@
...
-	struct bar *bar = (struct bar *)_bar;
-	dirty = handle_status_line(bar);
-}
-
-static void respond_output(int fd, short mask, void *_output) {
-	struct output *output = (struct output *)_output;
-	if (wl_display_dispatch(output->registry->display) == -1) {
-		sway_log(L_ERROR, "failed to dispatch wl: %d", errno);
+	ipc_get_workspaces(bar);
+	wl_list_for_each(output, &bar->outputs, link) {
```
Comment (emersion) [inline swaybar/bar.c]: `render_all_frames`

File: swaybar/render.c [hunk 11]
```diff
@@ -1,228 +1,120 @@
...
+static uint32_t render_binding_mode_indicator(cairo_t *cairo,
+		struct swaybar_config *config, const char *mode, double x,
+		uint32_t height) {
+	int text_width, text_height;
+	get_text_size(cairo, config->font, &text_width, &text_height,
+			1, true, "%s", mode);
+	uint32_t ideal_height = text_height + ws_vertical_padding * 2
+		+ border_width * 2;
+	if (height < ideal_height) {
+		height = ideal_height;
```
Comment (emersion) [inline swaybar/render.c]: `return ideal_height`?

**Pull request merged**